include/deal.II/multigrid/mg_smoother.h

```diff
@@ -997,13 +1094,86 @@ MGSmootherPrecondition<MatrixType, PreconditionerType, VectorType>::smooth
         {
           if (this->debug > 0)
             deallog << 'T';
-          if (i == 0 && u.all_zero())
-            *r = rhs;
-          else
-            {
-              matrices[level].Tvmult(*r,u);
-              r->sadd(-1.,1.,rhs);
-            }
+          matrices[level].Tvmult(*r,u);
```

Comment:
It depends on the application of course. On large scales, its cost is definitely not negligible, even though circumstances are rare that an `MPI_Allreduce` really is slower than a `vmult()` as long as the matrix has entries on all processors. But not of our coarse-level matrices have, so it will show up at some point in a nasty way.

The more important point is that with separate interfaces to the pre-smoothing (with zeroing) and post-smoothing (where the vector usually is not zero) the check will be false in 99% of cases. Thus, I do not see the point in doing an expensive operation any more. :-)